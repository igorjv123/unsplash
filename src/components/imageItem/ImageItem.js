import React from 'react';
import { connect } from 'react-redux';
import './style.sass';
import { Link, withRouter } from 'react-router-dom';
import setQuery from '../../logic/mainSearch/actions'
import {func, object} from "prop-types";

const ImageItem = (props) => {

    const handleTagClick = function(e) {
        props.searchByTag(e.target.id)
        props.history.push('/')
    }

    const { alt_description, user, urls, created_at, likes, tags, links, description } = props.activeImage;

    return(
        <div className='image-item'>
            <img className = 'image-item__image' src={urls.regular} alt={description || alt_description}/>

            <div className='image-item__info'>
                <p className='image-description'>{alt_description}</p>
                <p className='image-author'>author: {user.name}, likes: {likes}</p>
                <div className={'image-tags'}>
                    {tags.map(({ title }) =>
                        <button
                            className='image-tags__item'
                            onClick={handleTagClick}
                            key={`tags-list-${title}`}
                            id={title}
                        >
                            {title}
                        </button>
                    )}
                </div>
                <div className='image-item__buttons'>
                    <a target='_blank' download="image.jpg" href={links.download} className='image-button'>Download</a>
                    <Link className='image-button' to='/'>Back</Link>
                </div>

            </div>

        </div>
    )

}

ImageItem.propTypes = {
    activeImage: object,
    searchByTag: func
}

ImageItem.defaultProps = {
    activeImage: {
        alt_description: 'description',
        user: {
            name: 'John Doe'
        },
        urls: {},
        created_at: '01.01.01',
        likes: 1,
        tags: [],
        links: {}

    },
    searchByTag: () => {}
}

const mapStateToProps = (state) => {
    // const { activeImage } = state;
    // return { activeImage }
    return {}
}
const maoDispatchToProps = (dispatch) => {
    return {
        searchByTag(data) {
            dispatch(setQuery(data))
        }
    }
};

export default connect(mapStateToProps, maoDispatchToProps)(withRouter(ImageItem));

