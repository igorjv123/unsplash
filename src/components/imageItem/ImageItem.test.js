import React from 'react';
import { ImageItem } from "./ImageItem";
import { mount } from 'enzyme';

describe('ImageItem ', () => {
    const searchByTag = jest.fn();
    const historyMock = {
        push: jest.fn(),
        goBack: jest.fn()
    }
    const imageMock = {
        alt_description: 'alt_description',
        description: 'description',
        user: {
            name: 'John Doe'
        },
        urls: {
            regular: 'www.link.some'
        },
        likes: 1,
        tags: [
            {
                title: 'tagName'
            },
            {
                title: 'tagName2'
            }
        ],
        links: {
            download: 'link.download'
        }
    };

    beforeEach(function () {
        searchByTag.mockClear();
        historyMock.push.mockClear();
        historyMock.goBack.mockClear();
    });

    it('should render without crashing', () => {
        const wrapper = mount(<ImageItem activeImage={imageMock} searchByTag={searchByTag} history={historyMock}/>);

        expect(wrapper).toMatchSnapshot();
    });

    it('should call searchByTag() and history.push() after tag click ', () =>{
        const wrapper = mount(<ImageItem activeImage={imageMock} searchByTag={searchByTag} history={historyMock}/>)

        const tags = wrapper.find('.image-tags__item')
        tags.forEach((tag) => {
            tag.simulate('click')
        });

        expect(searchByTag).toHaveBeenCalledTimes(tags.length);
        expect(historyMock.push).toHaveBeenCalledTimes(tags.length);
    });

    it('should call history.goBack() on burron "back" click', () => {
        const wrapper = mount(<ImageItem activeImage={imageMock} searchByTag={searchByTag} history={historyMock}/>)

        const backButton = wrapper.find('.back-button')
        backButton.simulate('click');

        expect(historyMock.goBack).toHaveBeenCalled();
    })
});