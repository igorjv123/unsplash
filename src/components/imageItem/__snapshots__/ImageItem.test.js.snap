// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageItem  should render without crashing 1`] = `
<ImageItem
  activeImage={
    Object {
      "alt_description": "alt_description",
      "description": "description",
      "likes": 1,
      "links": Object {
        "download": "link.download",
      },
      "tags": Array [
        Object {
          "title": "tagName",
        },
        Object {
          "title": "tagName2",
        },
      ],
      "urls": Object {
        "regular": "www.link.some",
      },
      "user": Object {
        "name": "John Doe",
      },
    }
  }
  history={
    Object {
      "goBack": [MockFunction],
      "push": [MockFunction],
    }
  }
  searchByTag={[MockFunction]}
>
  <div
    className="image-item"
  >
    <img
      alt="description"
      className="image-item__image"
      src="www.link.some"
    />
    <div
      className="image-item__info"
    >
      <p
        className="image-description"
      >
        alt_description
      </p>
      <p
        className="image-author"
      >
        author: 
        John Doe
        , likes: 
        1
      </p>
      <div
        className="image-tags"
      >
        <button
          className="image-tags__item"
          id="tagName"
          key="tags-list-tagName"
          onClick={[Function]}
          tabIndex={0}
        >
          tagName
        </button>
        <button
          className="image-tags__item"
          id="tagName2"
          key="tags-list-tagName2"
          onClick={[Function]}
          tabIndex={0}
        >
          tagName2
        </button>
      </div>
      <div
        className="image-item__buttons"
      >
        <a
          className="image-button"
          download="image.jpg"
          href="link.download"
          tabIndex={0}
          target="_blank"
        >
          Download
        </a>
        <button
          className="image-button back-button"
          onClick={[Function]}
          tabIndex={0}
        >
          Back
        </button>
      </div>
    </div>
  </div>
</ImageItem>
`;
